dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/identitystore/InMemoryIdentityStoreDefinitionWrapper.java

```diff
@@ -0,0 +1,421 @@
+ *     }
+ * )
+ */
+
+public class InMemoryIdentityStoreDefinitionWrapper {
+
+    private static final TraceComponent tc = Tr.register(InMemoryIdentityStoreDefinitionWrapper.class);
+
+    /** The definitions for this IdentityStore. */
+    private final InMemoryIdentityStoreDefinition inMemoryIdentityStoreDefinition;
+
+    /**
+     * NOTE: For priority and useFor, we do not need to store priorityExpression, nor useForExpression as they
+     * effect the end values of priority and useFor respectively, but are not needed retrospectively.
+     */
+
+    /** The priority for this IdentityStore. Will be null when set by a deferred EL expression. */
+    private final Integer priority;
+
+    /** The ValidationTypes this IdentityStore can be used for. Will be null when set by a deferred EL expression. */
+    private final Set<ValidationType> useFor;
+
+    /** the annotations credentials in a key/value map: callerName = {password, groups[]} */
+    private final HashMap<String, CredentialValue> credentials;
+
+    private final ELHelperWrapper elHelper;
+
+    /**
+     * Create a new instance of an {@link InMemoryIdentityStoreDefinitionWrapper} that will provide
+     * convenience methods to access configuration from the {@link InMemoryIdentityStoreDefinition}
+     * instance.
+     *
+     * @param inMemoryIdentityStoreDefinition The {@link InMemoryIdentityStoreDefinition} to wrap.
+     */
+    public InMemoryIdentityStoreDefinitionWrapper(InMemoryIdentityStoreDefinition inMemoryIdentityStoreDefinition) {
+
+        if (inMemoryIdentityStoreDefinition == null) {
+            throw new IllegalArgumentException("The InMemoryIdentityStoreDefinition cannot be null.");
+        }
+        this.inMemoryIdentityStoreDefinition = inMemoryIdentityStoreDefinition;
+        elHelper = new ELHelperWrapper();
+
+        Tr.warning(tc, "JAKARTASEC_WARNING_PRODUCTION_USE");
+
+        /*
+         * Evaluate the configuration. The values will be non-null if the setting is NOT
+         * a deferred EL expression. If it is a deferred EL expression, we will dynamically
+         * evaluate it at call time.
+         */
+        priority = evaluatePriority(false);
+        useFor = evaluateUseFor(false);
+
+        // with the annotation from the application, parse out the values
+        int credentialsLength = (inMemoryIdentityStoreDefinition.value() == null) ? 0 : inMemoryIdentityStoreDefinition.value().length;
+        Credentials[] creds = Arrays.copyOf(inMemoryIdentityStoreDefinition.value(), credentialsLength);
+
+        // convert annotation credentials into an easily search-able HashMap with standard Java types
+        credentials = new HashMap<String, CredentialValue>(credentialsLength);
+        for (int i = 0; i < credentialsLength; i++) {
+            CredentialValue credential = getCredential(creds[i].callerName(), creds[i].password(), creds[i].groups());
+            credentials.put(creds[i].callerName(), credential);
+        }
```

Comment:
```suggestion
        Map<String, CredentialValue> credsMap = new HashMap<>(credentialsLength);
        for (int i = 0; i < credentialsLength; i++) {
            CredentialValue credential = getCredential(creds[i].callerName(), creds[i].password(), creds[i].groups());
            credsMap.put(creds[i].callerName(), credential);
        }
        credentials = Collections.unmodifiableMap(credsMap);
```
Of course you will need to change `credentials` type to `Map` instead of `HashMap`.

Reply:
Okay, a better programming practice to create an internal unmodifiable map instead of a modifiable one which relies on programming guards to ensure its immutability.  Nice suggestion, and done.